src/phoebus_driver.cpp

```diff
@@ -1014,11 +1017,211 @@ TaskListStatus PhoebusDriver::MonteCarloStep() {
 
 /**
  * Gets called before output.
- * Currently: Fills tracers.
+ * Currently:
+ *  Fills Tracers
+ *  Computes entropy for output
  **/
 void UserWorkBeforeOutput(MeshBlock *pmb, ParameterInput *pin) {
   auto tracer_pkg = pmb->packages.Get("tracers");
   bool do_tracers = tracer_pkg->Param<bool>("active");
+
+  namespace p = fluid_prim;
+  namespace diag = diagnostic_variables;
+  auto rc = pmb->meshblock_data.Get().get();
+  using parthenon::MakePackDescriptor;
+  Mesh *pmesh = rc->GetMeshPointer();
+  const int ndim = pmesh->ndim;
+  auto &resolved_pkgs = pmesh->resolved_packages;
+
+  static auto desc =
+      MakePackDescriptor<p::velocity, p::density, p::ye, p::temperature, p::entropy,
+                         p::cs, diag::ratio_divv_cs>(resolved_pkgs.get());
+  auto v = desc.GetPack(rc);
+  auto coords = pmb->coords;
+
+  auto geom = Geometry::GetCoordinateSystem(rc);
+
+  IndexRange ib = rc->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = rc->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = rc->GetBoundsK(IndexDomain::interior);
+
+  auto eos = pmb->packages.Get("eos")->Param<Microphysics::EOS::EOS>("d.EOS");
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "UserWorkBeforeOutput::H5", DevExecSpace(), kb.s, kb.e, jb.s,
+      jb.e, ib.s, ib.e, KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        Real lambda[2];
+        if (v.Contains(0, p::ye())) {
+          lambda[0] = v(0, p::ye(), k, j, i);
+        }
+        const Real s = eos.EntropyFromDensityTemperature(
+            v(0, p::density(), k, j, i), v(0, p::temperature(), k, j, i), lambda);
+        const Real p = eos.PressureFromDensityTemperature(
+            v(0, p::density(), k, j, i), v(0, p::temperature(), k, j, i), lambda);
+        const Real bmod = eos.BulkModulusFromDensityTemperature(
+            v(0, p::density(), k, j, i), v(0, p::temperature(), k, j, i), lambda);
+        const Real sie = eos.InternalEnergyFromDensityTemperature(
+            v(0, p::density(), k, j, i), v(0, p::temperature(), k, j, i), lambda);
+        const Real h = sie + p / v(0, p::density(), k, j, i) + 1;
+        const Real cs = std::sqrt(bmod / v(0, p::density(), k, j, i) / h);
+        Real divv;
+        Real gam[3][3];
+        Real gam1[3][3];
+        Real gam2[3][3];
+        Real gam3[3][3];
+        Real gam12[3][3];
+        Real gam13[3][3];
+        Real gam23[3][3];
+        Real gam123[3][3];
+        const Real vp[3] = {v(0, p::velocity(0), k, j, i), v(0, p::velocity(1), k, j, i),
+                            v(0, p::velocity(2), k, j, i)};
+        const Real vp1[3] = {v(0, p::velocity(0), k, j, i - 1),
+                             v(0, p::velocity(1), k, j, i - 1),
+                             v(0, p::velocity(2), k, j, i - 1)};
```

Comment:
I see you chose backward differences. I probably would have chosen centered but it doesn't matter much